pkg/ccl/utilccl/builtins.go

```diff
@@ -0,0 +1,214 @@
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/volatility"
+	"github.com/cockroachdb/cockroach/pkg/sql/types"
+)
+
+// RegisterableBuiltin must be a function whose argument types
+// are in BuiltinArgTypes and whose return type is in BuiltinReturnTypes.
+type RegisterableBuiltin interface{}
+
+// TODO: anyArg gets used at runtime, so for performance this should be
+// not be a reflect type, but that's tricky without using generics
+// or a potentially cumbersome union struct.
+type anyArg = reflect.Value
+
+func makeAnyArg(x interface{}) anyArg {
+	return reflect.ValueOf(x)
+}
+
+func union(v anyArg) interface{} {
+	return v.Interface()
+}
+
+func makePGError(v anyArg) error {
+	if v.IsNil() {
+		return nil
+	}
+	err := v.Interface().(error)
+	return pgerror.Wrap(err, pgerror.GetPGCode(err), "internal implementation error")
+}
+
+// NativeToDatum defines a conversion from a golang primitive or CCL
+// type into a tree.Datum.
+type NativeToDatum struct {
+	Type       *types.T
+	Conversion func(anyArg) tree.Datum
+}
+
+// DatumToNative defines a conversion from a tree.Datum to a golang
+// primitive or CCL type.
+type DatumToNative struct {
+	Type       *types.T
+	Conversion func(tree.Datum) anyArg
+}
+
+// BuiltinReturnTypes maps types returned by a go function to the equivalent
+// tree.Datum for a SQL function to return.
+var BuiltinReturnTypes = map[reflect.Type]NativeToDatum{
+	reflect.TypeOf("string"): {
+		Type:       types.String,
+		Conversion: func(x anyArg) tree.Datum { return tree.NewDString(union(x).(string)) },
+	},
+	reflect.TypeOf([]byte{}): {
+		Type:       types.Bytes,
+		Conversion: func(x anyArg) tree.Datum { return tree.NewDBytes(tree.DBytes(union(x).([]byte))) },
+	},
+}
+
+// BuiltinArgTypes maps argument types taken by a go function to an equivalent
+// tree.Datum for a SQL function to take.
+var BuiltinArgTypes = map[reflect.Type]DatumToNative{
+	reflect.TypeOf("string"): {
+		Type:       types.String,
+		Conversion: func(d tree.Datum) anyArg { return makeAnyArg(string(tree.MustBeDString(d))) },
+	},
+}
+
+// RegisterArgType allows a package-specific type to be used as input to a SQL function.
+func RegisterArgType(
+	zeroValue interface{}, datumType *types.T, converter func(tree.Datum) interface{},
+) {
+	BuiltinArgTypes[reflect.TypeOf(zeroValue)] = DatumToNative{
+		Type:       datumType,
+		Conversion: func(d tree.Datum) anyArg { return makeAnyArg(converter(d)) },
+	}
+}
+
+// RegisterCCLBuiltin converts an arbitrary go function to an equivalent SQL function
+// using reflection. Names should be namespaced (usually starting with `crdb_internal.ccl_`).
+// impl must be a function that returns a single type with a key in BuiltinReturnTypes,
+// plus optionally an error. This is a convenience method meant for generating SQL functions
+// that are tightly coupled to other behavior. It's generally preferable to create SQL builtins
+// via calling builtinsregistry.Register directly.
```

Comment:
I honestly don't know about this -- seems like too much magic.

As your comment says that it is "preferable to create and register via builtinsregistry"... why not make it required? What does this function add that builtinregistry doesn't provide in a very explicit way?